### PR TITLE
Dima/add extra account

### PR DESCRIPTION
**Pull request by dmitriy-matv**
Redux with multi-account activation

### CODE REVIEW

File: frontend/src/store/reducers/authReducer.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+// const localStorageUser = localStorage.getItem("currentUser");
+// let currentUser: CurrentUser = {};
+// if (localStorageUser) {
+//   currentUser = JSON.parse(localStorageUser);
+// }
+// console.log("log in auth reducer", currentUser);
+// const loggedIn =
+//   Object.keys(currentUser).length === 0 && currentUser.constructor === Object
+//     ? false
+//     : true;
```
Comment (istacat) [inline frontend/src/store/reducers/authReducer.ts:17]: don't store old code in commented fragments... use git :)

File: tests/test_auth.py [hunk 2]
```diff
@@ -59,3 +59,16 @@ def test_login_and_logout(client):
...
+    response = client.post(
+        "/api/accounts",
+        data=dict(
+            ad_login="QWE123",
+            ad_password="testTest",
+            created_by_id=1,
+            reseller_id=1,
+        ),
+    )
+    assert b"Registration successful. You are logged in." in response.data
```
Comment (istacat) [inline tests/test_auth.py:74]: please run flake8

**Review by istacat — COMMENTED** [review]
Thank you for the opportunity to make review for your code :)

**Review by kipod — APPROVED** [review]

**Review by Taraslysak — COMMENTED** [review]
Pretty nice)
Only one question. Why haven't u merged redux branch? Reading 80 files made my morning)

**Review by Taraslysak — APPROVED** [review]

**Comment by dmitriy-matv** [comment]
> Pretty nice)
> Only one question. Why haven't u merged redux branch? Reading 80 files made my morning)

I was waiting for kipod review))

File: frontend/src/components/forms/ExtraAccount.tsx [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  const [expiryDate, setExpiryDate] = useState<any>();
+
+  const [subscrPeriod, setSubscrPeriod] = useState<SubscriptionPeriod>(
+    SubscriptionPeriod.ONE_YEAR
+  );
+
+  const [sim, setSim] = useState("");
+  const [license, setLicense] = useState("");
+  const [imei, setImei] = useState("");
+  const [comment, setComment] = useState("");
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply frontend/src/components/forms/ExtraAccount.tsx:46]: You're right, already delete this)

File: frontend/src/components/forms/ExtraAccount.tsx [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    setExpiryDate(moment(expDate).format("MMM Do YYYY"));
+  };
+
+  useEffect(() => {
+    initAccountInstance();
+  }, []);
+
+  // useEffect(() => {
+
+  // }, []);
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply frontend/src/components/forms/ExtraAccount.tsx:84]: done)

File: frontend/src/components/forms/AddAccountForm.tsx [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    index: string,
+    attr: keyof IAccountFormState,
+    value: number | boolean | string | Date | [Date, Date] | null
+  ): void => {
+    const groupToUpdate = { ...createAccountFormsState[index] };
+    if (
+      attr === "sim" ||
+      attr === "imei" ||
+      attr === "comment" ||
+      attr === "license_key"
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply frontend/src/components/forms/AddAccountForm.tsx:51]: Fix with this code:

if (["sim", "imei", "comment", "license_key"].includes(attr)) { do staff }

File: frontend/src/components/forms/AddAccountForm.tsx [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            >
+              {resellersOptions}
+            </select>
+
+            <span className="form__input-error"></span>
+          </div>
+        </div>
+      </div>
+
+      {/* {forms} */}
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply frontend/src/components/forms/AddAccountForm.tsx:222]: done

File: frontend/src/components/forms/AddAccountForm.tsx [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+import "react-datepicker/dist/react-datepicker.css";
+import { AccountFetchService } from "../../helpers";
+import { useTypedSelector } from "../../helpers/useTypedSelector";
+import { accountFormInputs } from "../../types/accountTypes";
+import { useActions } from "../../helpers/useActions";
+import { ExtraAccount } from "./ExtraAccount";
+
+import { getInitialValues } from ".";
+import { IAccountFormState, validateForm } from "./FormikHelpers";
+import moment from "moment";
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply frontend/src/components/forms/AddAccountForm.tsx:15]: removed

File: frontend/src/App.tsx [hunk 8]
```diff
@@ -48,37 +72,39 @@ const App: React.FC = () => {
...
     <div className="splash-container">
       <p className="splash-element">
-        Please wait few seconds, we try to create new account.
+        Please wait few seconds, we are processing data.
       </p>
     </div>
   );
 
+  const errorComponent = () => {
+    const res = error.map((errorGroup) => {
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply frontend/src/App.tsx:90]: I'll try to fix it properly)) Thank's👌

File: frontend/src/App.tsx [hunk 9]
```diff
@@ -48,37 +72,39 @@ const App: React.FC = () => {
...
-  };
-  const logoutHandler = () => {
-    setLoggedIn(false);
-  };
-  const statusHandler = () => {
-    setSplashStatus((prev) => !prev);
-  };
   useEffect(() => {
     const currentPath = window.location.pathname.split("/")[1];
     const res = currentPath.charAt(0).toUpperCase() + currentPath.slice(1);
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply frontend/src/App.tsx:77]: 😊

File: app/views/api/subscription.py [hunk 10]
```diff
@@ -12,25 +13,23 @@ def __init__(self):
...
-        self.post_args.add_argument("subscr_id")
         self.post_args.add_argument("subscr_type")
         self.post_args.add_argument("subscr_period")
-        self.post_args.add_argument("activationDateController")
+        self.post_args.add_argument("activation_date")
         self.post_args.add_argument("ad_login")
         self.post_args.add_argument("Authorization", location="headers")
 
     def post(self):
+        log(log.DEBUG, "============== POST subscription begin ==========")
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply app/views/api/subscription.py:23]: done

File: app/views/api/account.py [hunk 11]
```diff
@@ -29,22 +30,29 @@ def __init__(self):
 
     def get(self):
         args = self.get_args.parse_args()
-        print(args)
+        log(log.INFO, "GET arguments %s", args)
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply app/views/api/account.py:33]: done

File: tests/test_auth.py (hunk 2 above)
Comment (dmitriy-matv) [reply tests/test_auth.py:74]: code was deleted)

File: frontend/src/store/reducers/authReducer.ts (hunk 1 above)
Comment (dmitriy-matv) [reply frontend/src/store/reducers/authReducer.ts:17]: done

File: frontend/src/components/forms/ExtraAccount.tsx [hunk 12]
```diff
@@ -0,0 +1,279 @@
+import React, { useEffect, useState } from "react";
+import DatePicker from "react-datepicker";
+import moment from "moment";
+import { IApiAnswer, SubscriptionPeriod } from "../../interfaces";
+import { AccountFetchService } from "../../helpers";
+import { IAccountFormState } from "./FormikHelpers";
+
+interface IAccList {
+  [x: string]: any;
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply frontend/src/components/forms/ExtraAccount.tsx:9]: removed)

File: app/views/api/subscription.py [hunk 13]
```diff
@@ -48,7 +47,7 @@ def post(self):
                     print(f"exception subscr {e}")
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply app/views/api/subscription.py:47]: done

File: app/views/api/account.py [hunk 14]
```diff
@@ -175,11 +195,16 @@ def get(self):
             isUnique = False
             while not isUnique:
                 rand_int = secrets.randbelow(17575999)
-                ecc_id = account.ecc_encode(rand_int)
-                acc = Account.query.filter_by(ecc_id=ecc_id).first()
+                ad_login = account.ecc_encode(rand_int)
+                acc = Account.query.filter_by(ad_login=ad_login).first()
                 if not acc:
                     isUnique = True
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply app/views/api/account.py:201]: done

File: app/views/api/account.py [hunk 15]
```diff
@@ -104,9 +115,18 @@ def get(self, account_id):
...
-                return account.to_json()
+                resp = {
+                    "status": "success",
+                    "message": "Account was created",
+                    "data": account.to_json(),
+                }
+                return resp
             else:
-                return {"error": "No such account"}
+                resp = {
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply app/views/api/account.py:125]: done

File: app/views/api/account.py [hunk 16]
```diff
@@ -104,9 +115,18 @@ def get(self, account_id):
         if status["status"] == "success":
             account = Account.query.get(account_id)
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply app/views/api/account.py:116]: done

File: app/views/api/account.py [hunk 17]
```diff
@@ -7,6 +7,7 @@
 # from app import db
 import secrets
 from app.controllers import account, mdm
+from app import log
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply app/views/api/account.py:10]: done

File: .gitignore [hunk 18]
```diff
@@ -126,3 +126,6 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+
+test.html
+test.css
```
[the remark this answers is not in the document]
Comment (dmitriy-matv) [reply .gitignore:131]: These files only for my internal usage, so we don't need them in repo

**Pull request merged**